clang/lib/DPCT/Schema.cpp

```diff
@@ -0,0 +1,335 @@
+    TS.TypeSize =
+        TS.FieldNum == 0
+            ? 0
+            : ceilDevide((TS.Members.back().Offset + TS.Members.back().ValSize),
+                         TS.TypeAlign);
+  }
+  STypeSchemaMap[KetStr] = TS;
+  return TS;
+}
+
+TypeSchema dpct::registerTypeSchema(const clang::QualType &QT) {
+  if (QT.isNull())
+    return TypeSchema();
+  const std::string &KetStr = DpctGlobalInfo::getTypeName(QT);
+  if (CTypeSchemaMap.find(KetStr) != CTypeSchemaMap.end())
+    return CTypeSchemaMap[KetStr];
+  clang::ASTContext &AstContext = dpct::DpctGlobalInfo::getContext();
+  TypeSchema TS;
+  TS.TypeName = KetStr;
+  TS.TypeSize = AstContext.getTypeSize(QT) / CHAR_BIT;
+  if (const clang::RecordType *RT = QT->getAs<clang::RecordType>()) {
+    TS.FileName =
+        getFilePathFromDecl(RT->getDecl(), DpctGlobalInfo::getSourceManager());
+    if (isa<clang::CXXRecordDecl>(RT->getDecl())) {
+      clang::CXXRecordDecl *RD = RT->getAsCXXRecordDecl();
+      // visit base class
+      DFSBaseClass(RD, TS);
+    }
+    for (const clang::FieldDecl *field : RT->getDecl()->fields()) {
+      TS.Members.push_back(constructFieldSchema(field));
+    }
+    TS.FieldNum = TS.Members.size();
+    TS.TypeAlign = AstContext.getTypeAlign(QT) / CHAR_BIT;
+  }
+  CTypeSchemaMap[KetStr] = TS;
+  registerSYCLTypeSchema(QT);
+  return TS;
+}
+
+VarSchema dpct::constructVarSchema(const clang::DeclRefExpr *DRE) {
+  clang::ASTContext &AstContext = dpct::DpctGlobalInfo::getContext();
+  VarSchema VA;
+  VA.VarName = DRE->getNameInfo().getAsString();
+  VA.IsBasicType =
+      getDerefType(DRE->getType().getCanonicalType())->isBuiltinType();
+  VA.ValTypeOfVar = getValType(DRE->getType());
+  VA.FileName =
+      getFilePathFromDecl(DRE->getDecl(), DpctGlobalInfo::getSourceManager());
+  VA.VarSize = AstContext.getTypeSize(DRE->getType()) / CHAR_BIT;
+  VA.VarType = DpctGlobalInfo::getTypeName(getDerefType(DRE->getType()));
+  VA.TypeSize = AstContext.getTypeSize(getDerefType(DRE->getType())) / CHAR_BIT;
+  if (!VA.IsBasicType) {
+    registerTypeSchema(getDerefType(DRE->getType()));
+  }
+  return VA;
+}
+
+VarSchema dpct::constructSyclVarSchema(const VarSchema &CVS) {
+  VarSchema VA = CVS;
+  // Is SYCL type
+  if (MapNames::TypeNamesMap.find(CVS.VarType) !=
+      MapNames::TypeNamesMap.end()) {
+    VA.VarType = MapNames::TypeNamesMap[CVS.VarType]->NewName;
+    // Can find SYCL Schema
+    if (STypeSchemaMap.find(VA.VarType) != STypeSchemaMap.end()) {
+      VA.TypeSize =
+          STypeSchemaMap[MapNames::TypeNamesMap[CVS.VarType]->NewName].TypeSize;
+      if (CVS.ValTypeOfVar == ValType::ArrayValue) {
+        VA.VarSize = CVS.VarSize / CVS.TypeSize * VA.TypeSize;
+      } else if (CVS.ValTypeOfVar == ValType::ScalarValue) {
+        VA.VarSize = VA.TypeSize;
+      }
+    }
+  }
+  return VA;
+}
+
+std::map<std::string, TypeSchema> clang::dpct::CTypeSchemaMap;
+
+std::map<std::string, TypeSchema> clang::dpct::STypeSchemaMap;
```

Comment:
For definition of  global variables, it is better to put the begin of this file, and add the comments for key and value.